src/core/net/srp_server.cpp

```diff
@@ -90,7 +90,6 @@ Server::Server(Instance &aInstance)
     , mOutstandingUpdatesTimer(aInstance, HandleOutstandingUpdatesTimer, this)
     , mEnabled(false)
 {
-    IgnoreError(SetDomain(kDefaultDomain));
```

Comment:
Maybe we can eliminate the necessity of `otHeapSetCAllocFree`, i.e. use platform method for external heap alloc and free, e.g. otPlatExternalHeapAlloc, otPlatExternalHeapFree

Reply:
I suggest in core module we continue to use `Instance::HeapAlloc()` (and not use platform layer `otPlat` directly). 

The intention behind having `Instanc::HeapAlloc()` is to provide a single API handling all different configurations (internal vs external, multiple/sinsgle instance, etc). Otherwise we need all `#if` check in the modules here.

Reply:
I am not suggesting to use `otPlatExternalHeapAlloc` method directly in core. 

Currently, the user (posix) code has to set up function pointers using `otHeapSetCAllocFree`. 
```cpp
#if OPENTHREAD_CONFIG_HEAP_EXTERNAL_ENABLE
    otHeapSetCAllocFree(calloc, free);
#endif
```

I am suggesting that we should define two more platform methods (i.e. otPlatExternalHeapAlloc and otPlatExternalHeapAFree) and platforms supporting external heap (e.g. POSIX) should define these two methods. Then, OT core is can freely call `Instance::HeapCAlloc` in constructors no matter it's using internal heap or external.

Reply:
I see. Got it. This makes sense to me and sounds good.

If we want to use heap APIs from constructors, one thing to be careful is to ensure `Heap` constructor itself is called before other object constructors. It seems to be high on the list of member variables in `instance.hpp` so we should be fine.

Reply:
> I am suggesting that we should define two more platform methods (i.e. otPlatExternalHeapAlloc and otPlatExternalHeapAFree) and platforms supporting external heap (e.g. POSIX) should define these two methods. Then, OT core is can freely call Instance::HeapCAlloc in constructors no matter it's using internal heap or external.

sounds good to me as well 👍 I think we can close this PR and create another for adding the external heap platform APIs.